src/gui/spatialArrow.js

```diff
@@ -0,0 +1,224 @@
+        canvasContainer.style.position = 'absolute';
+        canvasContainer.style.top = '0';
+        canvasContainer.style.left = '0';
+        document.body.appendChild(canvasContainer);
+
+        canvas = document.createElement('canvas');
+        canvas.className = 'arrow-canvas';
+        canvas.style.position = 'absolute';
+        menuBarHeight = realityEditor.device.environment.variables.screenTopOffset;
+        canvas.style.top = `${menuBarHeight}px`;
+        canvas.style.left = '0';
+        canvas.style.zIndex = '3001';
+        canvasContainer.appendChild(canvas);
+
+        ctx = canvas.getContext("2d");
+    }
+
+    function resizeCanvas() {
+        if (canvas !== undefined) {
+            screenW = window.innerWidth;
+            screenH = window.innerHeight - menuBarHeight;
+            screenRatio = screenH / screenW;
+            canvas.width = screenW;
+            canvas.height = screenH;
+        }
+    }
+
+    const Clamp = (x, low, high) => {
+        return Math.min(Math.max(x, low), high);
+    }
+
+    const Remap01 = (x, low, high) => {
+        return Clamp((x - low) / (high - low), 0, 1);
+    }
+
+    const Remap = (x, low1, high1, low2, high2) => {
+        return low2 + (high2 - low2) * Remap01(x, low1, high1);
+    }
+
+    let translateX = 0, translateY = 0;
+
+    function Translate(x, y) {
+        translateX += x;
+        translateY += y;
+        ctx.translate(x, y);
+    }
+
+    function Rotate(a) {
+        ctx.rotate(a);
+    }
+
+    function Scale(x, y) {
+        ctx.scale(x, y);
+    }
+
+    function clear() {
+        ctx.clearRect(-translateX, -translateY, screenW, screenH);
+    }
+
+    // draw an arrow at (x, y) center
+    function Arrow(x, y, rotation, scale, innerColor='rgb(0, 255, 255)', outerColor='rgb(255, 255, 255)') {
+        Translate(x, y);
+        Scale(scale, scale);
+        Rotate(rotation);
+        
+        // draw path
+        let region = new Path2D();
+        region.moveTo(0, -3);
+        region.lineTo(-2, 2);
+        region.lineTo(2, 2);
+        region.closePath();
+        // fill path
+        ctx.fillStyle = innerColor;
+        ctx.fill(region, 'evenodd');
+        // stroke path
+        ctx.strokeStyle = outerColor;
+        ctx.lineWidth = .4;
+        ctx.stroke(region);
+        
+        Rotate(-rotation);
+        Scale(1 / scale, 1 / scale);
+        Translate(-x, -y);
+    }
+
+    function initCanvas() {
+        // make (0, 0) the center of canvas
+        Translate(screenW / 2, screenH / 2);
+        clear();
+    }
+    
+    let indicators = [];
+    function searchForIndicators() {
+        // todo: auto detect the indicator names, instead of hard-coded 'cylinderIndicator'
+        indicators = realityEditor.gui.threejsScene.getObjectsByName('cylinderIndicator');
+    }
+    
+    let worldPos = new THREE.Vector3();
+    let screenX, screenY;
+    let screenBorderFactor = 0.97;
+    let desX = 0, desY = 0;
```

Comment:
These variables don't need to be global. It's good practice to declare them closer to their point of use when you can

Reply:
Good tip. Will keep them within local scope. Thx James!

Reply:
You're welcome! Great work throughout this PR!